File: Fusion/Scripts/Utility/Python/RCM Color Space Match.py
#!/usr/bin/env python
"""For DaVinci Resolve Color Grading"""
__author__ = "Michael<https://github.com/fukco>"
__version__ = "0.5.0"
__license__ = "MIT"

import logging
import os
import re

from get_resolve import get_bmd

metadata_parser = __import__("Metadata Parser")

gui_mode = 1


class ColorSpaceMatchRule:
    def __init__(self, manufacturer, gamma_notes, color_space_notes, input_color_space):
        self.manufacturer = manufacturer
        self.gamma_notes = gamma_notes
        self.color_space_notes = color_space_notes
        self.input_color_space = input_color_space


# create logger
logger = logging.getLogger("RCM_Color_Space_Match")
logger.setLevel(logging.DEBUG)

# create console handler and set level to debug
ch = logging.StreamHandler()
ch.setLevel(logging.DEBUG)

# create formatter
formatter = logging.Formatter('%(asctime)s - %(levelname)s - %(message)s')

# add formatter to ch
ch.setFormatter(formatter)

# add ch to logger
logger.addHandler(ch)

default_color_version_name = "Auto Generate Color Version"
color_space_match_list = [ColorSpaceMatchRule("Atomos", "CLog", "Cinema", "Canon Cinema Gamut/Canon Log"),
                          ColorSpaceMatchRule("Atomos", "CLog2", "Cinema", "Canon Cinema Gamut/Canon Log2"),
                          ColorSpaceMatchRule("Atomos", "CLog3", "Cinema", "Canon Cinema Gamut/Canon Log3"),
                          ColorSpaceMatchRule("Atomos", "F-Log", "F-Gamut", "FujiFilm F-Log"),
                          ColorSpaceMatchRule("Atomos", "V-Log", "V-Gamut", "Panasonic V-Gamut/V-Log"),
                          ColorSpaceMatchRule("Atomos", "SLog3", "SGamut3", "S-Gamut3/S-Log3"),
                          ColorSpaceMatchRule("Atomos", "SLog3", "SGamut3Cine", "S-Gamut3.Cine/S-Log3"),
                          ColorSpaceMatchRule("Atomos", "N-Log", "BT.2020", "Nikon N-Log"),
                          ColorSpaceMatchRule("Atomos", "HLG", "BT.2020", "Rec.2100 HLG"),

                          ColorSpaceMatchRule("Fujifilm", "F-log", "", "FujiFilm F-Log"),

                          ColorSpaceMatchRule("Panasonic", "V-Log", "V-Gamut", "Panasonic V-Gamut/V-Log"),

                          ColorSpaceMatchRule("Sony", "s-log2", "s-gamut", "S-Gamut/S-Log2"),
                          ColorSpaceMatchRule("Sony", "s-log3-cine", "s-gamut3-cine", "S-Gamut3.Cine/S-Log3"),
                          ColorSpaceMatchRule("Sony", "s-log3", "s-gamut3", "S-Gamut3/S-Log3"),
                          ]
color_space_match_map = {}
input_color_space_list = []
for item in color_space_match_list:
    color_space_match_map[(item.gamma_notes, item.color_space_notes)] = item.input_color_space
    if item.input_color_space not in input_color_space_list:
        input_color_space_list.append(item.input_color_space)

match_rules = {"rules": []}
manufacturers = []
for item in color_space_match_list:
    if item.manufacturer in manufacturers:
        index = manufacturers.index(item.manufacturer)
        match_rules["rules"][index]["details"].append(
            {"Gamma Notes": item.gamma_notes, "Color Space Notes": item.color_space_notes,
             "Input Color Space": item.input_color_space})
    else:
        manufacturers.append(item.manufacturer)
        match_rules["rules"].append({"manufacturer": item.manufacturer, "details": [
            {"Gamma Notes": item.gamma_notes, "Color Space Notes": item.color_space_notes,
             "Input Color Space": item.input_color_space}]})


def main_window():
    # some element IDs
    win_id = "com.xiaoli.RCMColorSpaceMatch"  # should be unique for single instancing
    tree_id = "MatchTree"
    gamma_notes = "Gamma Notes"
    color_space_notes = "Color Space Notes"
    input_color_space = "Input Color Space"

    # check for existing instance
    win = ui.FindWindow(win_id)
    if win:
        win.Show()
        win.Raise()
        exit()

    # define the window UI layout
    win = dispatcher.AddWindow({
        'ID': win_id,
        'Geometry': [300, 180, 800, 460],
        'WindowTitle': "RCM Color Space Match",
    },
        ui.VGroup([
            # color space match rule
            ui.Tree({"ID": tree_id}),

            ui.HGroup({"Weight": 0}, [
                ui.CheckBox(
                    {"ID": "EnableMetadataParser", "Text": "Enable Metadata Parser", "Weight": 0, "Checked": True}),
                ui.HGap(10),
                ui.CheckBox(
                    {"ID": "EnableDataLevelAdjustment", "Text": "Enable Assign Atomos Clips' Data Level", "Weight": 0}),
                ui.ComboBox({"ID": "DataLevelAdjustmentType", "Weight": 1})
            ]),

            ui.HGroup({"Weight": 0}, [
                ui.Button({"Text": "Match", "ID": "ExecuteButton", "Weight": 0}),
                ui.HGap(5),
                ui.Label({'Font': ui.Font({'Family': "Times New Roman"}), "ID": "InfoLabel"})
            ]),
        ])
    )

    def init_tree():
        items = win.GetItems()

        # Add a header row.
        hdr = items[tree_id].NewItem()
        hdr["Text"][0] = gamma_notes
        hdr["Text"][1] = color_space_notes
        hdr["Text"][2] = input_color_space
        items[tree_id].SetHeaderItem(hdr)

        # Number of columns in the Tree list
        items[tree_id]["ColumnCount"] = 3

        # Resize the Columns
        items[tree_id]["ColumnWidth"][0] = 200
        items[tree_id]["ColumnWidth"][1] = 200
        items[tree_id]["ColumnWidth"][2] = 360

    def init_combo():
        items = win.GetItems()

        items["DataLevelAdjustmentType"].AddItem('For Log and Legal Clips')
        items["DataLevelAdjustmentType"].AddItem('For All Clips')

    def show_message(message, t=0):
        if t == 0:
            win.GetItems()["InfoLabel"]["Text"] = f"<font color='#39CA41'>{message}</font>"
        elif t == 1:
            win.GetItems()["InfoLabel"]["Text"] = f"<font color='#922626'>{message}</font>"

    def load_color_space_match_rule():
        for manufacturerRule in match_rules["rules"]:
            items = win.GetItems()
            item = items[tree_id].NewItem()
            item["Text"][0] = manufacturerRule["manufacturer"]
            items[tree_id].AddTopLevelItem(item)
            for rule in manufacturerRule["details"]:
                item_child = items[tree_id].NewItem()
                item_child["Text"][0] = rule[gamma_notes]
                item_child["Text"][1] = rule[color_space_notes]
                item_child["Text"][2] = rule[input_color_space]
                item.AddChild(item_child)
            item["Expanded"] = True

    def click_execute_button(ev):
        logger.info("Start Processing.")
        show_message("Processing...")
        items = win.GetItems()
        if execute(assign_data_level_enabled=items["EnableDataLevelAdjustment"]["Checked"],
                   assign_type=items["DataLevelAdjustmentType"]["CurrentIndex"],
                   parse_metadata_enabled=items["EnableMetadataParser"]["Checked"]):
            show_message("All Down. Have Fun!")
        else:
            show_message("Some process failed, please check console log details.", 1)

    def close(ev):
        dispatcher.ExitLoop()

    init_tree()
    init_combo()
    load_color_space_match_rule()

    # assign event handlers
    win.On[win_id].Close = close
    win.On["ExecuteButton"].Clicked = click_execute_button
    win.Show()
    dispatcher.RunLoop()
    win.Hide()


def get_clips(folder, result):
    result.extend(folder.GetClipList())
    sub_folders = folder.GetSubFolders()
    for sub_folder in sub_folders.values():
        get_clips(sub_folder, result)


def assign_data_level(clip, metadata, assign_type):
    if metadata.get("Camera Manufacturer") == "Atomos":
        gamma_notes = metadata.get("Gamma Notes") if metadata.get("Gamma Notes") else ""
        camera_notes = metadata.get("Camera Notes") if metadata.get("Camera Notes") else ""
        if assign_type == 0 and (
                re.search("LOG", gamma_notes, re.IGNORECASE) or "Range: Legal" in camera_notes) or assign_type == 1:
            if clip.SetClipProperty("Data Level", "Full"):
                logger.debug(f"Assign {clip.GetName()} data level full successfully.")
            else:
                logger.error(f"Assign {clip.GetName()} data level full failed.")
                return False
    return True


def parse_metadata(clip, lib):
    file_path = clip.GetClipProperty("File Path")
    if len(file_path) > 0:
        resolve_meta_dict = lib.DRProcessMediaFile(file_path.encode("utf-8")).get_dict()
        if resolve_meta_dict:
            if not resolve_meta_dict["IsSupportMedia"]:
                logger.warning(f"{os.path.basename(file_path)} Not Supported.")
            else:
                del resolve_meta_dict["IsSupportMedia"]
                meta = {k: v.decode("utf-8") for k, v in resolve_meta_dict.items() if v}
                if not meta:
                    logger.debug(f"Ignore clip {os.path.basename(file_path)}.")
                else:
                    if clip.SetMetadata(meta):
                        logger.debug(f"Process {os.path.basename(file_path)} Successfully.")
                        return meta
                    else:
                        logger.error(f"Failed to set {os.path.basename(file_path)} metadata!")
        else:
            logger.error(f"Failed to parse clip {clip.GetName()}")
    return None


def execute(assign_data_level_enabled=True, assign_type=0, parse_metadata_enabled=True):
    logger.info("Start match input color space and apply custom grading rules.")
    resolve = bmd.scriptapp("Resolve")
    project_manager = resolve.GetProjectManager()
    project = project_manager.GetCurrentProject()
    media_pool = project.GetMediaPool()
    root_folder = media_pool.GetRootFolder()
    success = True
    is_rcm = "davinciYRGBColorManaged" in project.GetSetting("colorScienceMode")
    clips = []

    get_clips(root_folder, clips)
    lib = {}
    if parse_metadata_enabled:
        lib = metadata_parser.get_cdll_lib()
    for clip in clips:
        metadata = parse_metadata(clip, lib) if parse_metadata_enabled else clip.GetMetadata()
        if not metadata:
            continue
        if is_rcm:
            gamma_notes = metadata.get("Gamma Notes") if metadata.get("Gamma Notes") else ""
            color_space_rotes = metadata.get("Color Space Notes") if metadata.get("Color Space Notes") else ""
            input_color_space = color_space_match_map.get((gamma_notes, color_space_rotes))
            if input_color_space:
                if clip.SetClipProperty("Input Color Space", input_color_space):
                    logger.debug(f"{clip.GetName()} Set Input Color Space {input_color_space} Successfully.")
                else:
                    success = False
                    logger.error(f"{clip.GetName()} Set Input Color Space {input_color_space} Failed!")
            else:
                logger.warning(f"{clip.GetName()} Does Not Found Input Color Space Match Rule!")
        if assign_data_level_enabled:
            if not assign_data_level(clip, metadata, assign_type):
                success = False
    if success:
        logger.info("All Done, Have Fun!")
        return True
    else:
        logger.warning("Some error happened, please check console details.")
        return False


if __name__ == '__main__':
    bmd = get_bmd()
    if "gui_mode" in locals().keys() and gui_mode:
        fusion = bmd.scriptapp("Fusion")
        ui = fusion.UIManager
        dispatcher = bmd.UIDispatcher(ui)
        main_window()
    else:
        execute()

File: Fusion/Scripts/Utility/Lua/RCM Color Space Match.lua
gui_mode = 1

local color_space_match_list = {
    { manufacturer = "Atomos", details = {
        { gamma_notes = "CLog", color_science_mode = "Cinema", input_color_space = "Canon Cinema Gamut/Canon Log" },
        { gamma_notes = "CLog2", color_science_mode = "Cinema", input_color_space = "Canon Cinema Gamut/Canon Log2" },
        { gamma_notes = "CLog3", color_science_mode = "Cinema", input_color_space = "Canon Cinema Gamut/Canon Log3" },
        { gamma_notes = "F-Log", color_science_mode = "F-Gamut", input_color_space = "FujiFilm F-Log" },
        { gamma_notes = "V-Log", color_science_mode = "V-Gamut", input_color_space = "Panasonic V-Gamut/V-Log" },
        { gamma_notes = "SLog3", color_science_mode = "SGamut3", input_color_space = "S-Gamut3/S-Log3" },
        { gamma_notes = "SLog3", color_science_mode = "SGamut3Cine", input_color_space = "S-Gamut3.Cine/S-Log3" },
        { gamma_notes = "N-Log", color_science_mode = "BT.2020", input_color_space = "Nikon N-Log" },
        { gamma_notes = "HLG", color_science_mode = "BT.2020", input_color_space = "Rec.2100 HLG" }
    } },
    { manufacturer = "Fujifilm", details = {
        { gamma_notes = "F-log", color_science_mode = "", input_color_space = "FujiFilm F-Log" }
    } },
    { manufacturer = "Panasonic", details = {
        { gamma_notes = "V-Log", color_science_mode = "V-Gamut", input_color_space = "Panasonic V-Gamut/V-Log" }
    } },
    { manufacturer = "Sony", details = {
        { gamma_notes = "s-log2", color_science_mode = "s-gamut", input_color_space = "S-Gamut/S-Log2" },
        { gamma_notes = "s-log3-cine", color_science_mode = "s-gamut3-cine", input_color_space = "S-Gamut3.Cine/S-Log3" },
        { gamma_notes = "s-log3", color_science_mode = "s-gamut3", input_color_space = "S-Gamut3/S-Log3" },
        { gamma_notes = "rec2100-hlg", color_science_mode = "rec2020", input_color_space = "Rec.2100 HLG" },
    }
    }
}

local color_space_match_map = {}

for _, value in ipairs(color_space_match_list) do
    for _, detail in ipairs(value["details"]) do
        if color_space_match_map[detail["gamma_notes"]] then
            local exist = color_space_match_map[detail["gamma_notes"]]
            if exist[detail["color_science_mode"]] == nil then
                exist[detail["color_science_mode"]] = detail["input_color_space"]
            end
        else
            local child = {}
            child[detail["color_science_mode"]] = detail["input_color_space"]
            color_space_match_map[detail["gamma_notes"]] = child
        end
    end
end

local lib
if ffi.os == "Windows" then
    lib = ffi.load(fusion:MapPath('LuaModules:/resolve-metadata.dll'))
elseif ffi.os == "OSX" then
    if ffi.arch == "x64" then
        lib = ffi.load(fusion:MapPath('LuaModules:/resolve-metadata-amd64.dylib'))
    elseif ffi.arch == "arm64" then
        lib = ffi.load(fusion:MapPath('LuaModules:/resolve-metadata-arm64.dylib'))
    end
end

ffi.cdef [[
	typedef struct
	{
		bool IsSupportMedia;
		char *CameraType;
		char *CameraManufacturer;
		char *CameraSerial;
		char *CameraId;
		char *CameraNotes;
		char *CameraFormat;
		char *MediaType;
		char *TimeLapseInterval;
		char *CameraFps;
		char *ShutterType;
		char *Shutter;
		char *ISO;
		char *WhitePoint;
		char *WhiteBalanceTint;
		char *CameraFirmware;
		char *LensType;
		char *LensNumber;
		char *LensNotes;
		char *CameraApertureType;
		char *CameraAperture;
		char *FocalPoint;
		char *Distance;
		char *Filter;
		char *NDFilter;
		char *CompressionRatio;
		char *CodecBitrate;
		char *SensorAreaCaptured;
		char *PARNotes;
		char *AspectRatioNotes;
		char *GammaNotes;
		char *ColorSpaceNotes;
	} DRMetadata;
	extern __declspec(dllexport) DRMetadata DRProcessMediaFile(char* absPath);
]]

local function LogLine(message)
    print(string.format("[%s] %s", os.date("%Y-%m-%d %X"), message))
end

local function ProcessClip(clip)
    local filePath = clip:GetClipProperty("File Path")
    if filePath ~= "" then
        local c_str = ffi.new("char[?]", #filePath+1)
        ffi.copy(c_str, filePath)
        local res = lib.DRProcessMediaFile(c_str)
        if res == nil then
            LogLine("Failed to parse clip" .. filePath)
            return
        end
        if not res.IsSupportMedia then
            LogLine("Clip " .. filePath .. " Not Support.")
            return
        end
        local metadata = {}
        local returnVal = {}
        returnVal["Camera Type"] = ffi.string(res.CameraType)
        returnVal["Camera Manufacturer"] = ffi.string(res.CameraManufacturer)
        returnVal["Camera Serial #"] = ffi.string(res.CameraSerial)
        returnVal["Camera ID"] = ffi.string(res.CameraId)
        returnVal["Camera Notes"] = ffi.string(res.CameraNotes)
        returnVal["Camera Format"] = ffi.string(res.CameraFormat)
        returnVal["Media Type"] = ffi.string(res.MediaType)
        returnVal["Time-Lapse Interval"] = ffi.string(res.TimeLapseInterval)
        returnVal["Camera FPS"] = ffi.string(res.CameraFps)
        returnVal["Shutter Type"] = ffi.string(res.ShutterType)
        returnVal["Shutter"] = ffi.string(res.Shutter)
        returnVal["ISO"] = ffi.string(res.ISO)
        returnVal["White Point (Kelvin)"] = ffi.string(res.WhitePoint)
        returnVal["White Balance Tint"] = ffi.string(res.WhiteBalanceTint)
        returnVal["Camera Firmware"] = ffi.string(res.CameraFirmware)
        returnVal["Lens Type"] = ffi.string(res.LensType)
        returnVal["Lens Number"] = ffi.string(res.LensNumber)
        returnVal["Lens Notes"] = ffi.string(res.LensNotes)
        returnVal["Camera Aperture Type"] = ffi.string(res.CameraApertureType)
        returnVal["Camera Aperture"] = ffi.string(res.CameraAperture)
        returnVal["Focal Point (mm)"] = ffi.string(res.FocalPoint)
        returnVal["Distance"] = ffi.string(res.Distance)
        returnVal["Filter"] = ffi.string(res.Filter)
        returnVal["ND Filter"] = ffi.string(res.NDFilter)
        returnVal["Compression Ratio"] = ffi.string(res.CompressionRatio)
        returnVal["Codec Bitrate"] = ffi.string(res.CodecBitrate)
        returnVal["Sensor Area Captured"] = ffi.string(res.SensorAreaCaptured)
        returnVal["PAR Notes"] = ffi.string(res.PARNotes)
        returnVal["Aspect Ratio Notes"] = ffi.string(res.AspectRatioNotes)
        returnVal["Gamma Notes"] = ffi.string(res.GammaNotes)
        returnVal["Color Space Notes"] = ffi.string(res.ColorSpaceNotes)
        for key, value in pairs(returnVal) do
            if value ~= "" then
                metadata[key] = value
            end
        end
        if next(metadata) == nil then
            LogLine("Ignore Clip " .. filePath)
            return
        end
        if clip:SetMetadata(metadata) then
            LogLine("Process clip " .. filePath .. " successfully.")
            return true, metadata
        else
            LogLine("Failed to set clip " .. filePath .. " metadata")
        end
    end
end

local tree_id = "MatchTree"

local function GetClips(folder, result)
    for _, v in ipairs(folder:GetClipList()) do
        table.insert(result, v)
    end
    sub_folders = folder:GetSubFolders()
    for _, sub_folder in ipairs(sub_folders) do
        GetClips(sub_folder, result)
    end
end

local function AssignDataLevel(clip, metadata, assign_type)
    if metadata and metadata["Camera Manufacturer"] == "Atomos" then
        gamma_notes = metadata["Gamma Notes"] ~= nil and metadata["Gamma Notes"] or ""
        camera_notes = metadata["Camera Notes"] ~= nil and metadata["Camera Notes"] or ""
        if assign_type == 0 and (string.find(string.upper(gamma_notes), "LOG") or string.find(camera_notes, "Range: Legal")) or assign_type == 1 then
            if clip:SetClipProperty("Data Level", "Full") then
                LogLine(string.format("Assign %s data level full successfully.", clip:GetName()))
            else
                LogLine(string.format("Assign %s data level full failed.", clip:GetName()))
                return False
            end
        end
    end
    return true
end

local function ParseMetadata(clip)
    local result, metadata = ProcessClip(clip)
    if result then
        return metadata
    else
        return nil
    end
end

local function Execute(assign_data_level_enabled, assign_type, parse_metadata_enabled)
    if assign_data_level_enabled == nil then
        assign_data_level_enabled = true
    end
    if assign_type == nil then
        assign_type = 0
    end
    if parse_metadata_enabled == nil then
        parse_metadata_enabled = true
    end
    LogLine("Start match input color space and apply custom grading rules.")
    resolve = bmd.scriptapp("Resolve")
    project_manager = resolve:GetProjectManager()
    project = project_manager:GetCurrentProject()
    media_pool = project:GetMediaPool()
    root_folder = media_pool:GetRootFolder()
    local success = true
    is_rcm = string.find(project:GetSetting("colorScienceMode"), "davinciYRGBColorManaged")

    if not is_rcm then
        LogLine("RCM Not Enabled!")
        return false, "RCM Not Enabled!"
    end

    clips = {}

    GetClips(root_folder, clips)
    for _, clip in ipairs(clips) do
        metadata = parse_metadata_enabled and ParseMetadata(clip) or clip:GetMetadata()
        codec = clip:GetClipProperty('Video Codec')
        if not metadata or 'RED' == string.upper(codec) or string.find(string.upper(codec), 'RAW') then
            goto continue
        end
        if is_rcm then
            gamma_notes = metadata["Gamma Notes"]
            color_space_rotes = metadata["Color Space Notes"]
            if gamma_notes == nil then
                gamma_notes = ""
            end
            if not color_space_rotes then
                color_space_rotes = ""
            end
            if gamma_notes == "" and color_space_rotes == "" then
                goto continue
            end
            input_color_space = ""
            if color_space_match_map[gamma_notes] then
                input_color_space = color_space_match_map[gamma_notes][color_space_rotes]
            end
            if input_color_space ~= "" then
                if clip:GetClipProperty("Input Color Space") == input_color_space then
                    LogLine(string.format("%s Input Color Space is already set", clip:GetName()))
                    goto continue
                end
                if clip:SetClipProperty("Input Color Space", input_color_space) then
                    LogLine(string.format("%s Set Input Color Space %s Successfully.", clip:GetName(), input_color_space))
                else
                    LogLine(string.format("%s Set Input Color Space %s Failed!", clip:GetName(), input_color_space))
                end
            else
                success = false
                LogLine(string.format("%s Does Not Found Input Color Space Match Rule!", clip:GetName()))
            end
        end
        ::continue::
        if assign_data_level_enabled then
            if not AssignDataLevel(clip, metadata, assign_type) then
                success = false
            end
        end
    end
    if success then
        LogLine("All Done, Have Fun!")
        return true
    else
        LogLine("Some error happened, please check console details.")
        return false
    end
end

local function MainWindow()
    local ui = fu.UIManager
    local disp = bmd.UIDispatcher(ui)

    --define the window UI layout
    win = disp:AddWindow({
        ID = "RCMColorSpaceMatchWin",
        WindowTitle = "RCM Color Space Match",
        Spacing = 0,

        ui:VGroup {
            --color space match rule
            ui:Tree { ID = tree_id },

            ui:VGap(2),

            ui:HGroup {
                Weight = 0,
                ui:CheckBox { ID = "EnableMetadataParser", Text = "Enable Metadata Parser", Weight = 0, Checked = true },
                ui:CheckBox { ID = "EnableDataLevelAdjustment", Text = "Enable Assign Atomos Clips' Data Level", Weight = 0 },
                ui:ComboBox { ID = "DataLevelAdjustmentType", Weight = 1 }
            },

            ui:VGap(2),

            ui:HGroup {
                Weight = 0,
                ui:Button { Text = "Match", ID = "ExecuteButton", Weight = 0 },
                ui:HGap(5),
                ui:Label { Font = ui:Font { Family = "Times New Roman" }, ID = "InfoLabel" }
            }
        }
    })

    items = win:GetItems()

    --Add a header row.
    hdr = items[tree_id]:NewItem()
    hdr.Text[0] = "Gamma Notes"
    hdr.Text[1] = "Color Space Notes"
    hdr.Text[2] = "Input Color Space"
    items[tree_id]:SetHeaderItem(hdr)

    --Number of columns in the Tree list
    items[tree_id].ColumnCount = 3

    --Resize the Columns
    items[tree_id].ColumnWidth[0] = 200
    items[tree_id].ColumnWidth[1] = 200
    items[tree_id].ColumnWidth[2] = 260

    items.DataLevelAdjustmentType:AddItem('For Log and Legal Clips')
    items.DataLevelAdjustmentType:AddItem('For All Clips')

    for _, value in ipairs(color_space_match_list) do
        item = items[tree_id]:NewItem()
        item.Text[0] = value["manufacturer"]
        items[tree_id]:AddTopLevelItem(item)
        for _, detail in ipairs(value["details"]) do
            item_child = items[tree_id]:NewItem()
            item_child.Text[0] = detail["gamma_notes"]
            item_child.Text[1] = detail["color_science_mode"]
            item_child.Text[2] = detail["input_color_space"]
            item:AddChild(item_child)
        end
        item.Expanded = true
    end

    win:Resize({ 700, 480 });
    win:RecalcLayout();

    local function ShowMessage(message, t)
        t = t or 0
        if t == 0 then
            items.InfoLabel.Text = string.format("<font color='#39CA41'>%s</font>", message)
        elseif t == 1 then
            items.InfoLabel.Text = string.format("<font color='#922626'>%s</font>", message)
        end
    end

    function win.On.ExecuteButton.Clicked(ev)
        LogLine("Start Processing.")
        ShowMessage("Processing...")
        local success, message = Execute(items["EnableDataLevelAdjustment"]["Checked"],
                items["DataLevelAdjustmentType"]["CurrentIndex"],
                items["EnableMetadataParser"]["Checked"])
        if success then
            ShowMessage("All Down. Have Fun!")
        else
            if message then
                ShowMessage(message, 1)
            else
                ShowMessage("Some processes failed, please check console details.", 1)
            end
        end
    end

    -- The window was closed
    function win.On.RCMColorSpaceMatchWin.Close(ev)
        disp:ExitLoop()
    end

    win:Show()
    disp:RunLoop()
    win:Hide()
end

if gui_mode == 1 then
    MainWindow()
else
    Execute()
end